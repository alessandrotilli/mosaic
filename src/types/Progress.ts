import { IBase } from "./Base";

type ProgressType = "circular" | "linear";

export type ProgressProps = IBase & {
  type: ProgressType;
  withLabel?: boolean;
  color?: "primary" | "success" | "error";
  variant?: "determinate" | "indeterminate";
  value?: number;
};
