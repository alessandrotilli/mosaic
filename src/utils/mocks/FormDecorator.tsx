import { useCallback, useEffect, useState } from "react";
import { DecoratorFn } from "@storybook/react";

export type FormValue = boolean | number | string | null;

const FormDecorator: DecoratorFn = (Story, { args: { onChange: externalOnChange, value: externalValue, ...args } }) => {
  const [value, setValue] = useState(externalValue);

  useEffect(() => {
    if (externalValue !== value) {
      setValue(externalValue);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [externalValue]);

  const onChange = useCallback(
    (value: FormValue) => {
      externalOnChange && externalOnChange(value);
      setValue(value);
    },
    [externalOnChange]
  );

  return Story({
    args: {
      ...args,
      onChange,
      value,
    },
  });
};

export default FormDecorator;
