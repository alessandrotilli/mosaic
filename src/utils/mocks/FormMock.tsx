import React, { cloneElement, FC, PropsWithChildren, ReactElement, useMemo, useState } from "react";
import { DecoratorFn } from "@storybook/react";

type IFormValue = boolean | number | string | null;

interface FormMockType {
  onChangePropName?: string;
  value: IFormValue;
  valuePropName?: string;
}

const FormMock: FC<PropsWithChildren<FormMockType>> = ({
  children,
  onChangePropName = "onChange",
  value: externalValue,
  valuePropName = "value",
}) => {
  const [value, setValue] = useState(externalValue);

  const wrappedFormElement = useMemo(
    () =>
      cloneElement(children as ReactElement<any>, {
        [onChangePropName]: setValue,
        [valuePropName]: value,
      }),
    [children, onChangePropName, value, valuePropName]
  );

  return wrappedFormElement;
};

const formDecorator: DecoratorFn = (Story, { args }) => <FormMock value={args.value}>{Story()}</FormMock>;

export { formDecorator, FormMock };
