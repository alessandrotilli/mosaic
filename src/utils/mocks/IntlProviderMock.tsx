import React, { FC } from "react";
import { IntlProvider } from "react-intl";

export enum LocaleMock {
  en = "en",
  it = "it-IT",
}

export enum MessageMock {
  button = "button",
  cancel = "cancel",
  checkbox = "checkbox",
  confirm = "confirm",
  inputNumber = "inputNumber",
  inputText = "inputText",
  subtitle = "subtitle",
  switch = "switch",
  title = "title",
  typography = "typography",
}

interface IntlProviderMockType {
  locale?: LocaleMock;
}

export const mockedMessages = {
  [LocaleMock.en]: {
    [MessageMock.button]: "Button",
    [MessageMock.cancel]: "Cancel",
    [MessageMock.checkbox]: "Checkbox",
    [MessageMock.confirm]: "Confirm",
    [MessageMock.inputNumber]: "Input Number",
    [MessageMock.inputText]: "Input Text",
    [MessageMock.subtitle]: "Subtitle",
    [MessageMock.switch]: "Switch",
    [MessageMock.title]: "Title",
    [MessageMock.typography]: "Typography",
  },
  [LocaleMock.it]: {
    [MessageMock.button]: "Bottone",
    [MessageMock.cancel]: "Annulla",
    [MessageMock.checkbox]: "Spunta",
    [MessageMock.confirm]: "Conferma",
    [MessageMock.inputNumber]: "Numero in input",
    [MessageMock.inputText]: "Testo in input",
    [MessageMock.subtitle]: "Sottotitolo",
    [MessageMock.switch]: "Interruttore",
    [MessageMock.title]: "Titolo",
    [MessageMock.typography]: "Tipografia",
  },
};

const IntlProviderMock: FC<IntlProviderMockType> = ({ children, locale = LocaleMock.en }) => {
  return (
    <IntlProvider locale={locale} messages={mockedMessages[locale]}>
      {children}
    </IntlProvider>
  );
};

export default IntlProviderMock;
