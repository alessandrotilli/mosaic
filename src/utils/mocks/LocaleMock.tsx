import React from "react";
import { DecoratorFn } from "@storybook/react";

import { LocalizedContextProvider } from "../../contexts/Localized";

export type ILocaleMock = "en" | "it-IT";

export enum MessageMock {
  button = "intl.message.button",
  cancel = "intl.message.cancel",
  checkbox = "intl.message.checkbox",
  confirm = "intl.message.confirm",
  inputNumber = "intl.message.inputNumber",
  inputText = "intl.message.inputText",
  subtitle = "intl.message.subtitle",
  switch = "intl.message.switch",
  title = "intl.message.title",
  typography = "typography",
}

type IMessagesMock = {
  [key in ILocaleMock]: {
    [key in MessageMock]: string;
  };
};

export const MESSAGES_MOCK: IMessagesMock = {
  en: {
    [MessageMock.button]: "Button",
    [MessageMock.cancel]: "Cancel",
    [MessageMock.checkbox]: "Checkbox",
    [MessageMock.confirm]: "Confirm",
    [MessageMock.inputNumber]: "Input Number",
    [MessageMock.inputText]: "Input Text",
    [MessageMock.subtitle]: "Subtitle",
    [MessageMock.switch]: "Switch",
    [MessageMock.title]: "Title",
    [MessageMock.typography]: "Typography",
  },
  "it-IT": {
    [MessageMock.button]: "Bottone",
    [MessageMock.cancel]: "Annulla",
    [MessageMock.checkbox]: "Spunta",
    [MessageMock.confirm]: "Conferma",
    [MessageMock.inputNumber]: "Numero in input",
    [MessageMock.inputText]: "Testo in input",
    [MessageMock.subtitle]: "Sottotitolo",
    [MessageMock.switch]: "Interruttore",
    [MessageMock.title]: "Titolo",
    [MessageMock.typography]: "Tipografia",
  },
};

const getLocalizedMessage = (key: MessageMock, locale: ILocaleMock): string => MESSAGES_MOCK[locale][key];

const localeDecorator: DecoratorFn = (Story) => (
  <LocalizedContextProvider localize={(key) => getLocalizedMessage(key as MessageMock, "it-IT")}>
    {Story()}
  </LocalizedContextProvider>
);

export { getLocalizedMessage, localeDecorator };
