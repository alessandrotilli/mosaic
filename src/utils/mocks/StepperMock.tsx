/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { cloneElement, FC, PropsWithChildren, ReactElement, useCallback, useEffect, useMemo } from "react";
import { DecoratorFn } from "@storybook/react";

interface StepperMockType {
  stepList?: [];
  activeStep?: number;
}

const StepperMock: FC<PropsWithChildren<StepperMockType>> = ({
  children,
  stepList,
  activeStep: externalActiveStep,
}) => {
  const [activeStep, setActiveStep] = React.useState(externalActiveStep || 0);

  useEffect(() => {
    if (externalActiveStep && stepList) {
      if (externalActiveStep > stepList?.length - 1 || externalActiveStep < 0) {
        setActiveStep(0);
      }
    } else {
      setActiveStep(externalActiveStep || 0);
    }
  }, [externalActiveStep, stepList]);

  const handleNext = useCallback(() => {
    setActiveStep((prevActiveStep) => prevActiveStep + 1);
  }, []);

  const handleBack = useCallback(() => {
    setActiveStep((prevActiveStep) => prevActiveStep - 1);
  }, []);

  const wrappedStepper = useMemo(
    () =>
      cloneElement(children as ReactElement<any>, {
        onBackClick: handleBack,
        onNextClick: handleNext,
        activeStep,
      }),
    [children, handleBack, handleNext, activeStep]
  );

  return <div>{wrappedStepper}</div>;
};

const stepperDecorator: DecoratorFn = (Story, { args }) => (
  <StepperMock activeStep={args.activeStep} stepList={args.stepList}>
    {Story()}
  </StepperMock>
);

export { stepperDecorator, StepperMock };
