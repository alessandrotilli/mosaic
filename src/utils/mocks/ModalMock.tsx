/* eslint-disable @typescript-eslint/no-explicit-any */
import React, {
  cloneElement,
  FC,
  PropsWithChildren,
  ReactElement,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";
import { DecoratorFn } from "@storybook/react";

import Button from "../../components/Button";
import { Icons } from "../../types/Icon";

interface ModalMockType {
  buttonLabel?: string;
  open?: boolean;
}

const ModalMock: FC<PropsWithChildren<ModalMockType>> = ({
  buttonLabel = "Open",
  children,
  open: externalOpen = false,
}) => {
  const [open, setOpen] = useState(externalOpen);

  useEffect(() => {
    if (externalOpen !== open) {
      setOpen(externalOpen);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [externalOpen]);

  const toggleModal = useCallback(() => setOpen((prevOpen) => !prevOpen), [setOpen]);

  const wrappedModal = useMemo(
    () =>
      cloneElement(children as ReactElement<any>, {
        closable: true,
        onClose: toggleModal,
        open,
      }),
    [children, open, toggleModal]
  );

  return (
    <div>
      <Button icon={{ name: Icons.open_new }} label={buttonLabel} onClick={toggleModal} variant="outlined" />
      {wrappedModal}
    </div>
  );
};

const modalDecorator: DecoratorFn = (Story) => <ModalMock>{Story()}</ModalMock>;

export { modalDecorator, ModalMock };
