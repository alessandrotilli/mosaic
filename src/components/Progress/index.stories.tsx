import { Meta, StoryObj } from "@storybook/react";

import { localeDecorator } from "../../utils/mocks/LocaleMock";
import getDocsPage from "../../utils/stories";

import Progress, { DATA_CY_DEFAULT } from ".";

const COMPONENT_NAME = "Progress";
Progress.displayName = COMPONENT_NAME;

const meta = {
  title: "Feedback/Progress",
  component: Progress,
  decorators: [localeDecorator],
  parameters: {
    docs: {
      ...getDocsPage({
        basedOn: {
          label: "MUI Progress Component",
          url: "https://mui.com/material-ui/react-progress/",
        },
        component: COMPONENT_NAME,
        e2eTestInfo: {
          dataCyDefault: DATA_CY_DEFAULT,
        },
      }),
    },
  },
} satisfies Meta<typeof Progress>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Circular: Story = {
  args: {
    type: "circular",
  },
};

export const CircularColor: Story = {
  args: {
    ...Circular.args,
    color: "error",
  },
};

export const CircularDeterminate: Story = {
  args: {
    ...Circular.args,
    variant: "determinate",
    value: 50,
  },
};

export const CircularInvalidValue: Story = {
  args: {
    ...Circular.args,
    value: -999,
  },
};

export const CircularWithLabel: Story = {
  args: {
    ...CircularDeterminate.args,
    withLabel: true,
  },
};

export const Linear: Story = {
  args: {
    type: "linear",
  },
};

export const LinearColor: Story = {
  args: {
    ...Linear.args,
    color: "error",
  },
};

export const LinearDeterminate: Story = {
  args: {
    ...Linear.args,
    variant: "determinate",
    value: 50,
  },
};

export const LinearInvalidValue: Story = {
  args: {
    ...Linear.args,
    value: 999,
  },
};

export const LinearWithLabel: Story = {
  args: {
    ...LinearDeterminate.args,
    withLabel: true,
  },
};
