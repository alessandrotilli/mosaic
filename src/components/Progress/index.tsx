import React, { FC, useMemo } from "react";
import { Box, LinearProgress as MUILinearProgress } from "@mui/material";
import MUICircularProgress from "@mui/material/CircularProgress";
import Typography from "@mui/material/Typography";

import { ProgressType } from "../../types/Progress";

export const DATA_CY_DEFAULT = "progress";

const Progress: FC<ProgressType> = ({
  dataCy = DATA_CY_DEFAULT,
  color = "primary",
  type = "Circular",
  withLabel,
  variant,
  value = 100,
}) => {
  const percent = useMemo(() => {
    if (value < 0) {
      return 0;
    } else if (value > 100) {
      return 100;
    }
    return value;
  }, [value]);

  if (type === "Circular") {
    return (
      <Box sx={{ position: "relative", display: "inline-flex" }}>
        <MUICircularProgress data-cy={dataCy} color={color} variant={variant} value={percent} />
        {withLabel && (
          <Box
            sx={{
              top: 0,
              left: 0,
              bottom: 0,
              right: 0,
              position: "absolute",
              display: "flex",
              alignItems: "center",
              justifyContent: "center",
            }}
          >
            <Typography variant="caption" component="div" color="text.secondary">
              {percent + `%`}
            </Typography>
          </Box>
        )}
      </Box>
    );
  }
  return (
    <Box sx={{ display: "flex", alignItems: "center" }}>
      <Box sx={{ width: "100%", mr: 1 }}>
        <MUILinearProgress data-cy={dataCy} color={color} variant={variant} value={percent} />
      </Box>
      {withLabel && (
        <Box sx={{ minWidth: 35 }}>
          <Typography variant="body2" color="text.secondary">
            {percent + `%`}
          </Typography>
        </Box>
      )}
    </Box>
  );
};

export default Progress;
