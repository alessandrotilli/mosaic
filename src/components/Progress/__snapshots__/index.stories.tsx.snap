// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback/Progress Circular smoke-test 1`] = `
<div style="display: inline-flex; position: relative;">
  <span class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        data-cy="progress"
        style="width: 40px; height: 40px;"
  >
    <svg class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
         viewbox="22 22 44 44"
    >
      <circle class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              r="20.2"
              fill="none"
              stroke-width="3.6"
      >
      </circle>
    </svg>
  </span>
</div>
`;

exports[`Feedback/Progress CircularColor smoke-test 1`] = `
<div style="display: inline-flex; position: relative;">
  <span class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorError css-1lxih1g-MuiCircularProgress-root"
        role="progressbar"
        data-cy="progress"
        style="width: 40px; height: 40px;"
  >
    <svg class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
         viewbox="22 22 44 44"
    >
      <circle class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              r="20.2"
              fill="none"
              stroke-width="3.6"
      >
      </circle>
    </svg>
  </span>
</div>
`;

exports[`Feedback/Progress CircularDeterminate smoke-test 1`] = `
<div style="display: inline-flex; position: relative;">
  <span class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorPrimary css-1rn30mb-MuiCircularProgress-root"
        role="progressbar"
        aria-valuenow="50"
        data-cy="progress"
        style="width: 40px; height: 40px; transform: rotate(-90deg);"
  >
    <svg class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
         viewbox="22 22 44 44"
    >
      <circle class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              r="20.2"
              fill="none"
              stroke-width="3.6"
              style="stroke-dasharray: 126.92; stroke-dashoffset: 63.46px;"
      >
      </circle>
    </svg>
  </span>
</div>
`;

exports[`Feedback/Progress CircularInvalidValue smoke-test 1`] = `
<div style="display: inline-flex; position: relative;">
  <span class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        data-cy="progress"
        style="width: 40px; height: 40px;"
  >
    <svg class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
         viewbox="22 22 44 44"
    >
      <circle class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              r="20.2"
              fill="none"
              stroke-width="3.6"
      >
      </circle>
    </svg>
  </span>
</div>
`;

exports[`Feedback/Progress CircularWithLabel smoke-test 1`] = `
<div style="display: inline-flex; position: relative;">
  <span class="MuiCircularProgress-root MuiCircularProgress-determinate MuiCircularProgress-colorPrimary css-1rn30mb-MuiCircularProgress-root"
        role="progressbar"
        aria-valuenow="50"
        data-cy="progress"
        style="width: 40px; height: 40px; transform: rotate(-90deg);"
  >
    <svg class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
         viewbox="22 22 44 44"
    >
      <circle class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate css-oxts8u-MuiCircularProgress-circle"
              cx="44"
              cy="44"
              r="20.2"
              fill="none"
              stroke-width="3.6"
              style="stroke-dasharray: 126.92; stroke-dashoffset: 63.46px;"
      >
      </circle>
    </svg>
  </span>
  <div class="MuiTypography-root MuiTypography-caption css-ol7wxs-MuiTypography-root"
       style="align-items: center; display: flex; justify-content: center; position: absolute; inset: 0px;"
  >
    50%
  </div>
</div>
`;

exports[`Feedback/Progress Linear smoke-test 1`] = `
<div style="align-items: center; display: flex;">
  <span class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
        data-cy="progress"
        style="margin-right: 8px; width: 100%;"
  >
    <span class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-5ir5xx-MuiLinearProgress-bar1">
    </span>
    <span class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-1r8wrcl-MuiLinearProgress-bar2">
    </span>
  </span>
</div>
`;

exports[`Feedback/Progress LinearColor smoke-test 1`] = `
<div style="align-items: center; display: flex;">
  <span class="MuiLinearProgress-root MuiLinearProgress-colorError MuiLinearProgress-indeterminate css-1qzfjjf-MuiLinearProgress-root"
        role="progressbar"
        data-cy="progress"
        style="margin-right: 8px; width: 100%;"
  >
    <span class="MuiLinearProgress-bar MuiLinearProgress-barColorError MuiLinearProgress-bar1Indeterminate css-trpfwy-MuiLinearProgress-bar1">
    </span>
    <span class="MuiLinearProgress-bar MuiLinearProgress-barColorError MuiLinearProgress-bar2Indeterminate css-1ta3urv-MuiLinearProgress-bar2">
    </span>
  </span>
</div>
`;

exports[`Feedback/Progress LinearDeterminate smoke-test 1`] = `
<div style="align-items: center; display: flex;">
  <span class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
        aria-valuenow="50"
        aria-valuemin="0"
        aria-valuemax="100"
        data-cy="progress"
        style="margin-right: 8px; width: 100%;"
  >
    <span class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(-50%);"
    >
    </span>
  </span>
</div>
`;

exports[`Feedback/Progress LinearInvalidValue smoke-test 1`] = `
<div style="align-items: center; display: flex;">
  <span class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
        data-cy="progress"
        style="margin-right: 8px; width: 100%;"
  >
    <span class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-5ir5xx-MuiLinearProgress-bar1">
    </span>
    <span class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-1r8wrcl-MuiLinearProgress-bar2">
    </span>
  </span>
</div>
`;

exports[`Feedback/Progress LinearWithLabel smoke-test 1`] = `
<div style="align-items: center; display: flex;">
  <span class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-determinate css-eglki6-MuiLinearProgress-root"
        role="progressbar"
        aria-valuenow="50"
        aria-valuemin="0"
        aria-valuemax="100"
        data-cy="progress"
        style="margin-right: 8px; width: 100%;"
  >
    <span class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Determinate css-5xe99f-MuiLinearProgress-bar1"
          style="transform: translateX(-50%);"
    >
    </span>
  </span>
  <p class="MuiTypography-root MuiTypography-body2 css-r40f8v-MuiTypography-root"
     style="min-width: 40px;"
  >
    50%
  </p>
</div>
`;
